Calculate pedestals in ndlar_flow

sam-fogarty opened the pull request:
I have implemented a workflow to calculate channel by channel pedestals. The code is based on the [original script](https://github.com/larpix/larpix-v2-testing-scripts/blob/master/event-display/gen_pedestal_json.py) for generating the pedestal json. The workflow is meant to be run on its own, separate from the main charge workflow. It produces both an hdf5 file with channel pedestals as well as the normal json. A pixel unique ID LUT is implemented which `calib_prompt_hits` and `generate_pedestals.py` uses (so there should no longer be any discrepancies of how those are calculated).

I plotted the pedestal values from 2x2 with the `ndlar_flow` method along with the old script. It shows pedestals from 
1) the ndlar_flow version
2) the larpix-v2-testing-scripts version (the default json used in ndlar_flow, made by someone else)
3) the larpix-v2-testing-scripts version that I generated with same configuration as 1)

Configuration used:
- vref_dac: 223
- vcm_dac: 68
- adc_counts: 256
- vdda: 1800
- mean_trunc: 3 # ADCs

File used: `packet-cold-pedestal-2024_06_05_08_28_19_CDT.h5`
So 1) and 3) match perfectly which tells me the two scripts generate the same pedestal values. They differ from 2) likely because I am not using the same configuration used to generate the original 2x2 pedestal json.
![2x2_pedestals](https://github.com/user-attachments/assets/c621026e-c81e-4f43-8995-f6d1477bfe1f)

For FSD I used the following configuration:
- vref_dac: 223
- vcm_dac: 85
- adc_counts: 255
- vdda: 1720
- mean_trunc: 3 # ADCs

Pedestal file used: `packet-pedestal-2024_11_12_03_08_18_CET.h5`
Self-trigger file used: `packet-0020095-2024_11_05_22_04_36_CET.h5`
The results for FSD look pretty good, a lot better than what I was getting before. Question: Does this look correct? 
Tagging some people who may have input, also I would appreciate reviews of the code: @krwood @jaafar-chakrani @russellphysics @seg188 
<img width="858" alt="FSD_hits" src="https://github.com/user-attachments/assets/0fa362fa-1023-449d-b17a-26215b165efb" />

Pedestals for FSD, comparing old pedestals (from `/global/common/software/dune/inputs/FSD/pedestals/packet-pedestal-2024_11_12_03_29_26_CET.pedestal.json`) to new pedestals (from `/global/cfs/cdirs/dune/users/sfogarty/flow_LowEnergy/FSD/pedestal_tests/packet-pedestal-2024_11_12_03_08_18_CET_ped.json`)
<img width="861" alt="pedestals_FSD" src="https://github.com/user-attachments/assets/5216afe9-2a9d-4449-94a4-33f84533dc91" />

sam-fogarty committed:
Add LUT for pixel unique id

sam-fogarty committed:
Use pixel unique id LUT, use input vref/vcm as defaults in dictionaries

sam-fogarty committed:
Add workflow for generating channel pedestals

diaza commented on src/proto_nd_flow/reco/charge/calib_prompt_hits.py:
```diff
@@ -244,10 +250,7 @@ def run(self, source_name, source_slice, cache):
-                            + tile_id.astype(int)*100_000
-                            + packets_arr['chip_id'].astype(int)*100
-                            + packets_arr['channel_id'].astype(int))
+            hit_uniqueid = resources['Geometry'].pixel_unique_id[(packets_arr['io_group'], tile_id, packets_arr['chip_id'], packets_arr['channel_id'])]
```
So before a hit_uniqueid was always constructed, while in the current implementation hit_uniqueid would be -1 if the pixel channel wasn't defined in the pixel layout file. Do we know the behavior of this? Does it just use all default values downstream?

diaza commented:
It looks good to me. 
@jaafar-chakrani had made a script that calculates the thresholds [here](https://github.com/DUNE/larnd-sim/blob/develop/examples/Thesholds.ipynb). It might be good to move these scripts into the same place, or at least document well where to find them.

diaza commented:
Actually, could you maybe add a blurb somewhere (maybe the readme?) that says something like "here is where you can generate the pedestal files"?

sam-fogarty commented:
Thanks @diaza for taking a look and providing some comments. Upon rethinking this, I think it is an unnecessary step to make the pixel_unique_id LUT, or at least I don't see what it gives us (maybe a slight performance boost?). In principle the geometry will contain all the channels that exist in the data, but I worry sometimes this might not be the case, so the question is how we would handle the cases where the channel is not defined in the geometry? I think what we should do instead is effectively do what we did before, where the pedestal step will generate pedestals, calculate unique ids, and store them in a json file. Then, in `calib_prompt_hits`, we take the data, calculate unique ids, and look up the pedestals for those channels in the pedestal json. It is much less likely I think that the unique id will not be found in the pedestal json, since they are both data (with the same pixels, and presumably pretty similar CRO configuration)  and the unique ids are calculated in the exact same way. Rather than the available unique ids being generated from the geometry, which may not contain all the channels from data. In this case I would probably move the unique ids function to another file, maybe in `util` (I want to avoid putting it in `calib_prompt_hits` just in case another scripts need to use it, so it should be put in a common place).

I can also put some reference to this in the README as you suggest.

diaza commented:
I think that modification makes sense, as it'll simplify the code and I don't think anything would be lost. 
I think we will get hits that aren't in the pedestal file due to corrupted data, so we just have to make sure that things don't crash then.

krwood reviewed (approved):
Looks good, thanks Sam. I only have one organizational question. Why put it in `utils` subdir? Given this is run on charge data, maybe we should put it in the `charge` subdir?

sam-fogarty commented:
Hi @krwood thanks for taking a look. I can definitely put it in the charge subdir, it makes more sense that way.

krwood commented:
@russellphysics @jaafar-chakrani @seg188 Any comments on the 2x2 pedestal differences? I think FSD is understood, but can you check the configurations that were used to take the 2x2 pedestals and confirm which file is correct?

sam-fogarty committed:
Remove unique id LUT, replaced with unique id function. Moved from util subdir to charge subdir.

krwood commented on src/proto_nd_flow/reco/charge/pixel_functions.py:
```diff
@@ -0,0 +1,39 @@
+    :param adc_counts: nominally 2^N, where N is the total bits in the ADC (usually 8)
+    :returns: array of pedestal-corrected datawords converted to mV units
+    '''
+    return (vref-vcm) * adc/adc_counts + cm
```
typo: `cm` --> `vcm`

krwood commented on src/proto_nd_flow/reco/charge/pixel_functions.py:
Even though this one is also strictly crs - it could go into utils, given it's just some functions to use in other flow routines, and not a flow routine itself.

sam-fogarty replied:
I have moved that python file to util, now.

sam-fogarty committed:
Fix typos

sam-fogarty committed:
Move pixel_functions to util subdir

sam-fogarty committed:
Optimize pedestal generation script

sam-fogarty committed:
Update README

sam-fogarty commented:
Hi @krwood @diaza , I have made the following changes:
- Optimized pedestal generation script (much faster now)
- Move pixel_functions to `util` since it is not part of a normal workflow 
- Fixed a few typos
- Updated `README` with reference to the pedestal generation
- Unique id is no longer a LUT and is just a function that scripts can call, this is to avoid using the geometry to create the available unique ids (in case there are some unique ids existing in data but not in the geometry)

I would appreciate if you could take another look at the code with the new changes.

russellphysics commented:
@sam-fogarty 

There may be two possible discrepancies:

ADC counts == [255](https://github.com/DUNE/ndlar_flow/blob/746accf1a6fedb5a5f5c74d2006d1dec5756be57/yamls/fsd_flow/reco/charge/GeneratePedestals.yaml#L11) (instead of 256 as used in the [2x2 yaml](https://github.com/DUNE/ndlar_flow/blob/746accf1a6fedb5a5f5c74d2006d1dec5756be57/yamls/proto_nd_flow/reco/charge/GeneratePedestals.yaml#L11))

single file used of [14 files available](https://portal.nersc.gov/project/dune/data/FSD/nearline/packet/CRS/pedestal_11Nov2024/PRC_256/); we took pedestal data over 40 minutes at a slower trigger rate; we need to confirm that the pedestal means produced are representative of the full stats channel distribution


These issues may not be problematic but it is worthwhile checking

krwood commented:
@sam-fogarty : @jaafar-chakrani just confirmed the FSD data was taken with `vref_dac = 223`, `vcm_dac = 68`. Could you remake the FSD pedestal file with those settings and using more data files? We should at least check how many pedestal samples per channel we're getting in the file you used.

sam-fogarty commented:
Thanks Kevin. I will look at addressing this this week.

———————————
Samuel Fogarty
Graduate Research Assistant
Colorado State University
Department of Physics
AR124 Engineering Building, 400 Isotope Dr, Fort Collins, CO 80523
***@***.***
________________________________
From: Kevin Wood ***@***.***>
Sent: Tuesday, July 1, 2025 1:53:12 PM
To: DUNE/ndlar_flow ***@***.***>
Cc: Fogarty,Sam ***@***.***>; Mention ***@***.***>
Subject: Re: [DUNE/ndlar_flow] Calculate pedestals in ndlar_flow (PR #196)


** Caution: EXTERNAL Sender **

[https://avatars.githubusercontent.com/u/18149958?s=20&v=4]krwood left a comment (DUNE/ndlar_flow#196)<https://github.com/DUNE/ndlar_flow/pull/196#issuecomment-3024999466>

@sam-fogarty<https://github.com/sam-fogarty> : @jaafar-chakrani<https://github.com/jaafar-chakrani> just confirmed the FSD data was taken with vref_dac = 223, vcm_dac = 68. Could you remake the FSD pedestal file with those settings and using more data files? We should at least check how many pedestal samples per channel we're getting in the file you used.

—
Reply to this email directly, view it on GitHub<https://github.com/DUNE/ndlar_flow/pull/196#issuecomment-3024999466>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/APLFC7FFXK4A44ABJECRY433GLDIRAVCNFSM6AAAAAB7HBJNHOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTAMRUHE4TSNBWGY>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

sam-fogarty commented:
See new PR: https://github.com/DUNE/ndlar_flow/pull/202
